### PR TITLE
gh-143008: Fix Null pointer dereferences in TextIOWrapper underlying stream access

### PR DESCRIPTION
**Pull request by cmaloney**
TextIOWrapper keeps its underlying stream in a member called `self->buffer`. That stream can be detached by user code, such as custom `.flush()` implementations, resulting in `self->buffer` being set to NULL. The implementation often checked at the start of functions if `self->buffer` is in a good state but did not always recheck after other Python code which could modify `self->buffer`.

The cases which need to be re-checked are hard to spot so rather than rely on reviewer effort make a better safety net by changing all `self->buffer` access to go through helper functions.

Thank you @yihong0618 for the test, NEWS and an initial implementation in gh-143041.

* Issue: gh-143008

### CODE REVIEW

File: Lib/test/test_io/test_textio.py [hunk 1]
```diff
@@ -1565,8 +1565,7 @@ def closed(self):
         wrapper.close()  # should not crash
 
     def test_reentrant_detach_during_flush(self):
-        # gh-143008: Reentrant detach() during flush should raise RuntimeError
-        # instead of crashing.
+        # gh-143008: Reentrant detach() during flush should not crash.
```
Comment (yihong0618) [inline Lib/test/test_io/test_textio.py:1564]: cool seems better

Comment (cmaloney) [reply Lib/test/test_io/test_textio.py:1564]: Could you clarify what the goal / what you want with this comment? Is it more that you'd prefer RuntimeError or some similar piece? Would you like me to not use and update the test you wrote but build a separate one?

Overall trying to get this to a place it is hopefully shippable before the 3.15 beta.

Comment (yihong0618) [reply Lib/test/test_io/test_textio.py:1564]: no just agree this sorry for misleading it

**Comment by cmaloney** [comment]
@colesbury could you take a look as you've done some recent work on TextIO safety around threading. These cases are generally single threaded but generally the pattern "Implementation usually checked state; then did an operation which could modify it and didn't recheck".

**Comment by colesbury** [comment]
I'm a bit confused about the relationship here between `_textiowrapper_buffer_safe` and `CHECK_ATTACHED`. When can `buffer` be `NULL` and `CHECK_ATTACHED` not trigger?

**Comment by colesbury** [comment]
Less important, but please match the [CPython style](https://peps.python.org/pep-0007/) for C code, e.g.:

* `{` on new lines for functions definitions
* when wrapping function calls, indent continuation lines to align with the call site

**Comment by cmaloney** [comment]
> I'm a bit confused about the relationship here between _textiowrapper_buffer_safe and CHECK_ATTACHED. When can buffer be NULL and CHECK_ATTACHED not trigger?

`CHECK_ATTACHED`, by way of `CHECK_INITIALIZED`, checks `self->ok` first. The `self->ok` flag is set to `0` at start of `__init__`, `1` at the end, and cleared/reset to `0` during `.clear()` as well as at the start of the `Py_tp_dealloc`. During `__init__` there are calls to check the file is in a good state and get some attributes (remember: `self->buf` is the underlying file-like object not an internal buffer). 

The more complicated case is during deletion where there is an interaction with the base class. `TextIOWrapper` implements `Py_tp_dealloc` and its base class (`TextIOBase` which has a base class `IOBase`) implements a `Py_tp_finalize` with the function `iobase_finalize`. That function tries to ensure all data is written in the full I/O "stack" before any part of it is closed out of order. It tries to have Text I/O flush and close, then Buffered I/O (`self->buf`), then the Raw I/O. If there is an underlying file (`self->buf != NULL`) during those operations need to try and write data in the `TextIOWrapper` even though `self->ok` is `0`.

Adding to the complexity here all the objects involved and the multiple operations (`flush()`, `close()`, `.closed`) can be overridden and could call `TextIOWrapper.clear()` or `TextIOWrapper.detach()` causing `self->buf` to become `NULL`.

> Less important, but please match the [CPython style](https://peps.python.org/pep-0007/) for C code, e.g.:

Sorry for missing those bits, doing a review of that and will update for it shortly.

**Comment by cmaloney** [comment]
There are a couple cases I'm not sure how to PEP 7, in particular:
```c
/* hits 81 cols */
        PyObject *bytes = _textiowrapper_buffer_callmethod_noargs(self,
                                                                  &_Py_ID(read));
/* hits 81 cols */
            res = _textiowrapper_buffer_callmethod_onearg(self,
                                                          &_Py_ID(_dealloc_warn),
                                                          (PyObject *)self);


PyObject *input_chunk = _textiowrapper_buffer_callmethod_onearg(self,
            &_Py_ID(read), bytes_to_feed);
```

A couple directions I looked at but don't know how to decide between:
1. Make the variable name shorter (increases diff)
2. Make the function name shorter (ex. `_textiowrapper_buf_...` or `_textio_buffer_`, `_textiowrapper_buf_callmeth_noargs`)
3. Whats the right indentation to just put on the next line?

**Comment by colesbury** [comment]
I think it's fine to make the function names shorter, i.e., `buffer_callmethod_onearg` instead of `_textiowrapper_buffer_callmethod_onearg`. File-scope static functions don't need a prefix or underscore.

In general the preference is for wrapping function calls like the following (from PEP 7 example):

```c
PyErr_Format(PyExc_TypeError,
             "cannot create '%.100s' instances",
             type->tp_name);
```

But if that's not practical due to overly long lines, than just try to fit the style of surround code. For example, from elsewhere in `textio.c`:

```c
        PyObject *incrementalDecoder = PyObject_CallFunctionObjArgs(
            (PyObject *)state->PyIncrementalNewlineDecoder_Type,
            self->decoder, self->readtranslate ? Py_True : Py_False, NULL);
```

**Comment by cmaloney** [comment]
This should also resolve gh-143007

**Comment by cmaloney** [comment]
I think thish is ready for re-review @colesbury

File: Modules/_io/textio.c [hunk 2]
```diff
@@ -729,6 +731,63 @@ struct textio
...
+            PyErr_SetString(PyExc_ValueError,
+                            "I/O operation on uninitialized object");
+        }
+        else {
+            PyErr_SetString(PyExc_ValueError,
+                            "underlying buffer has been detached");
+        }
+        return NULL;
+    }
+    return self->buffer;
```
Comment (vstinner) [inline Modules/_io/textio.c:759]: I was worried that the function returns a borrowed reference. Another thread can call detach() converting the buffer variable to a dangling pointer. I hacked the code to introduce sleep+`sched_yield()` between getting the buffer and using the buffer, and I spawned 250 threads calling stream.write() and 1 thread calling stream.detach() with a random sleep.

In fact, it's (currently) safe to use a borrowed reference because all io.TextIOWrapper methods are protected by `@critical_section`. So it's not possible to call deatch() while another io.TextIOWrapper method is called: method calls are serialized by the critical section.

But it would be interesting to add a comment to explain why it's safe to use a borrowed reference. Something like:

```
// Returning a borrowed reference is safe since TextIOWrapper
// methods are protected by critical sections.
```

File: Modules/_io/textio.c [hunk 3]
```diff
@@ -729,6 +731,63 @@ struct textio
 
 #define textio_CAST(op) ((textio *)(op))
 
+/* Helpers to safely operate on self->buffer.
+
+   self->buffer can be detached (set to NULL) by any user code that is called
+   leading to NULL pointer dereferences (see gh-143008, gh-142594). Protect against
+   that by using helpers to check self->buffer validity at callsites. */
+static PyObject *
+buffer_access_safe(textio *self)
```
Comment (vstinner) [inline Modules/_io/textio.c:740]: I suggest to rename buffer_xxx() functions to textiowrapper_buffer_xxx(), and rename this function to `textiowrapper_buffer_get()`.

File: Modules/_io/textio.c [hunk 4]
```diff
@@ -729,6 +731,63 @@ struct textio
...
+
+static PyObject *
+buffer_getattr(textio *self, PyObject *attr_name)
+{
+    PyObject *buffer = buffer_access_safe(self);
+
+    if (buffer == NULL) {
+        return NULL;
+    }
+    return PyObject_GetAttr(buffer, attr_name);
```
Comment (vstinner) [inline Modules/_io/textio.c:770]: Nitpick: you may move the NULL check just after buffer_access_safe():

```suggestion
    PyObject *buffer = buffer_access_safe(self);
    if (buffer == NULL) {
        return NULL;
    }

    return PyObject_GetAttr(buffer, attr_name);
```

File: Modules/_io/textio.c [hunk 5]
```diff
@@ -3255,8 +3321,12 @@ static PyObject *
 _io_TextIOWrapper_closed_get_impl(textio *self)
 /*[clinic end generated code: output=b49b68f443a85e3c input=7dfcf43f63c7003d]*/
 {
+    /* During destruction self->ok is 0 but self->buffer is non-NULL and this
+       needs to error in that case which the safe buffer wrapper does not.
+
+       Match original behavior by calling CHECK_ATTACHED explicitly. */
```
Comment (vstinner) [inline Modules/_io/textio.c]: textiowrapper_dealloc() sets ok to 0 and almost immediately sets buffer to NULL. So I don't understand well when it's possible that ok is 0 and buffer is non-NULL.

This comment is unclear to me. I suggest rephrasing it to something like:

"Call CHECK_ATTACHED() to raise an exception if ..., when ... happens."

The "Match original behavior" part is unclear to me.

File: Modules/_io/textio.c (hunk 3 above)
Comment (cmaloney) [reply Modules/_io/textio.c:740]: Started like that (see: https://github.com/python/cpython/pull/145957/commits/d7f14fc87152d86ba5c6b478550f3dbfacdbc559) but it made some of the PEP-7 formatting really bad so moved to a shorter name

File: Modules/_io/textio.c (hunk 2 above)
Comment (cmaloney) [reply Modules/_io/textio.c:759]: Added a `_Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED` to validate + comment to document.

Comment (cmaloney) [reply Modules/_io/textio.c:759]: and found out `__init__` wasn't in a critical section in the process so re-initalization was racy... fixed as well

File: Modules/_io/textio.c (hunk 5 above)
Comment (cmaloney) [reply Modules/_io/textio.c]: Moved to checking `self->ok` directly and just returning true in that case which differs more in this bug fix but I think is a lot more understandable. Described a couple cases this is hit by the test suite.

File: Lib/test/test_io/test_textio.py [hunk 6]
```diff
@@ -1560,6 +1564,53 @@ def closed(self):
...
+        tests = [
+            ('truncate', lambda: wrapper.truncate(0)),
+            ('close', lambda: wrapper.close()),
+            ('detach', lambda: wrapper.detach()),
+            ('seek', lambda: wrapper.seek(0)),
+            ('tell', lambda: wrapper.tell()),
+            ('reconfigure', lambda: wrapper.reconfigure(line_buffering=True)),
+        ]
+        for name, method in tests:
+            with self.subTest(name), set_recursion_limit(100):
```
Comment (vstinner) [inline Lib/test/test_io/test_textio.py]: Why do you reduce the recursion limit?

File: Lib/test/test_io/test_textio.py [hunk 7]
```diff
@@ -1560,6 +1564,53 @@ def closed(self):
...
+                wrapper = self.TextIOWrapper(EvilBuffer(self.MockRawIO()), encoding='utf-8')
+                wrapper_ref = weakref.ref(wrapper)
+                # Used to crash; now will run out of stack.
+                self.assertRaises(RecursionError, method)
+                wrapper_ref = lambda: None
+                del wrapper
```
Comment (vstinner) [inline Lib/test/test_io/test_textio.py]: Would it work to just assign *wrapper* to `None` (and leave *wrapper_ref* unchanged)?

File: Modules/_io/textio.c [hunk 8]
```diff
@@ -672,6 +672,8 @@ struct textio
     int ok; /* initialized? */
     int detached;
     Py_ssize_t chunk_size;
+    /* Use helpers buffer_* to access buffer; many operations can set it to
```
Comment (vstinner) [inline Modules/_io/textio.c]:
```suggestion
    /* Use helpers buffer_*() functions to access buffer; many operations can set it to
```

File: Modules/_io/textio.c [hunk 9]
```diff
@@ -3255,8 +3326,17 @@ static PyObject *
 _io_TextIOWrapper_closed_get_impl(textio *self)
 /*[clinic end generated code: output=b49b68f443a85e3c input=7dfcf43f63c7003d]*/
 {
-    CHECK_ATTACHED(self);
-    return PyObject_GetAttr(self->buffer, &_Py_ID(closed));
+    /* If partially constructed or deconstructed return that the underlying
```
Comment (vstinner) [inline Modules/_io/textio.c]:
```suggestion
    /* If partially constructed or deconstructed, return that the underlying
```

**Review by vstinner — APPROVED** [review]
LGTM.

I don't think that it's worth it to backport the "check buffer" fixes. https://github.com/python/cpython/issues/143008 issue was found by fuzzing. Regular Python code should never meet this bug in practice. Triggering the code requires to design a very precise class which misbehaves.

But IMO the critical section on `TextIOWrapper.__init__` should be backported. Once this PR is merged, you can create a fix for 3.15 (and then we can backport it).

File: Lib/test/test_io/test_textio.py [hunk 10]
```diff
@@ -11,7 +11,11 @@
 import weakref
 from collections import UserList
 from test import support
-from test.support import os_helper, threading_helper
+from test.support import (
```
Comment (serhiy-storchaka) [inline Lib/test/test_io/test_textio.py]: It is better to not mix import of module attributes and submodules in one statement.

Complex multiline import is less readable than separate oneline imports.

File: Lib/test/test_io/test_textio.py (hunk 6 above)
Comment (serhiy-storchaka) [reply Lib/test/test_io/test_textio.py]: Why use unlimited recursion at all? Call the method, detach the buffer, set the flag, call the method yet one time if the flag was not set.

**Review by serhiy-storchaka — COMMENTED** [review]
I wonder, would not it be simpler if set `buffer` to None instead of NULL? How does it work in the Python implementation?

**Comment by cmaloney** [comment]
`_pyio` uses `None` to indicate no buffer. Can definitely migrate `_io.TextIOWrapper` to do that but that's a substantial code delta for an edge case and doesn't simplify fixing the time of check time of use issue here. Moving to None would make it calling methods on None rather than a null pointer de-reference. Would still need most this change of "always check _just_ before using".

re: Recursion most the tests don't rely on it. The one that does is when `flush()` calls `detach` itself which results in another call to `flush()`. gh-143041 resolved that by making a new "state" effectively of "are detaching, don't detach again". Can do a similar pattern here if want to eliminate the recursion potentially: In `.detach()` change `self->detached` and/or `self->ok` _before_ calling `_PyFile_Flush(self)` and making sure `.flush()` handles that case appropriately. I'm hesitant to do that as it changes the set of states a `TextIOWrapper` can be in a bit and I don't have those well mapped out mentally. Limiting recursion means the test exits after some number of recursion instances which validates it no longer segfaults / NULL pointer dereference is resolved.

File: Lib/test/test_io/test_textio.py (hunk 7 above)
Comment (cmaloney) [reply Lib/test/test_io/test_textio.py]: Removed wrapper_ref entirely and reworked how the classes work to simplify

File: Lib/test/test_io/test_textio.py (hunk 10 above)
Comment (cmaloney) [reply Lib/test/test_io/test_textio.py]: Removed set_recursion_limit entirely

File: Lib/test/test_io/test_textio.py (hunk 6 above)
Comment (cmaloney) [reply Lib/test/test_io/test_textio.py]: Removed the recursion

**Review by cmaloney — COMMENTED** [review]
Found a refactoring that removed the recursion while not making too much complex

**Comment by vstinner** [comment]
Merged, thanks @cmaloney for the fix.

No one should meet this bug in practice, since it requires a custom buffer object which detach() when the stream is not supposed to be detached. But well, it's good to fix the code to support this corner case.

@cmaloney: You can create a change for the 3.15 branch to backport the critical section if you want.

**Comment by cmaloney** [comment]
Critical section only backport: gh-151203
